rust/relay/server/src/main.rs

```diff
@@ -483,7 +483,12 @@ where
         loop {
             let mut ready = false;
 
-            ready!(self.sockets.flush(cx))?;
+            // Don't use `ready!` here - we must not block the loop on socket flushing.
+            // If we do, the phoenix channel never gets polled and heartbeats don't get sent,
+            // causing the portal to close our connection.
+            if let Poll::Ready(result) = self.sockets.flush(cx) {
```

Comment:
The problem with this approach is that you are breaking back-pressure. Unless the sockets are flushed, we should not queue any more work.